### PR TITLE
Add `as_ptr` and `as_mut_ptr` inherent method to `String`

### PR DESCRIPTION
**Pull request by Noratrieb**
Before, they went through `&str` and `&mut str`, which created intermediary references, shrinking provenance to only the initialized parts. `Vec<T>` already has such inherent methods added in #61114.

`beef` ran into this in [beef#47](https://github.com/maciejhirsz/beef/pull/47). 

The docs are mostly copied from `Vec::{as_ptr, as_mut_ptr}`, adapting the examples to something better fitting. The implementation simply forwards to the vec methods.

I'm not entirely sure what the correct stability attributes are for this, but I think it needs to be insta-stable like the vec methods?

### CODE REVIEW

**Comment by Nilstrieb** [comment]
r? rust-lang/libs-api @rustbot label +T-libs-api -T-libs

**Comment by Nilstrieb** [comment]
And I just found another case where this would have been helpful, [compart_str#100](https://github.com/ParkMyCar/compact_str/pull/100)

**Comment by Nilstrieb** [comment]
@RalfJung you might be interested as well, since you added the method to `Vec<T>`

File: library/alloc/src/string.rs [hunk 1]
```diff
@@ -1612,6 +1612,81 @@ impl String {
...
 
+    /// Returns a raw pointer to the string's buffer.
+    ///
+    /// The caller must ensure that the string outlives the pointer this
+    /// function returns, or else it will end up pointing to garbage.
+    /// Modifying the string may cause its buffer to be reallocated,
+    /// which would also make any pointers to it invalid.
+    ///
+    /// The caller must also ensure that the memory the pointer (non-transitively) points to
+    /// is never written to (except inside an `UnsafeCell`) using this pointer or any pointer
```
Comment (RalfJung) [inline library/alloc/src/string.rs:1623]:
```suggestion
    /// is never written to using this pointer or any pointer
```
There is no `UnsafeCell` here.

**Review by RalfJung — COMMENTED** [review]
Makes sense to me. However, since this changes the libs API surface, I cannot approve it -- let's wait for the @rust-lang/libs-api team.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting. Without suggesting that we have any consensus on the implied provenance handling here, we don't have objections to merging this unstably for now. However, since it doesn't sound like there's a way to do that yet, marking this as blocked.

**Comment by joshtriplett** [comment]
(Elaborating on the "implied provenance handling": these and the underlying `&str`/`&mut str` methods would return the same values here, and questions about provenance handling differences between them have not been settled.)

**Comment by saethlin** [comment]
The opening comment understates the impact that this has on the API. I'm hoping that this helps explain all the other implications of implementing `as_ptr` and `as_mut_ptr` without going through slice Deref:
```rust
fn main() {
    let mut s: Vec<u8> = Vec::with_capacity(10);
    s.extend(b"hello");
    let ptr = s.as_ptr(); // This pointer stays valid...
    let other = s.as_mut_ptr(); // Across as as_mut_ptr()
    let v = s; // Across a move of the original container
    unsafe {
        *other = b'b'; // And across a write to an aliasing pointer
        dbg!(*ptr);
    }

    let mut s = String::with_capacity(10);
    s.push_str("hello");
    let ptr = s.as_ptr(); // This pointer...
    let other = s.as_mut_ptr(); // Is invalidated here (though this case might be fixed, UCG 113)
    let v = s; // But wouldn't be invalidated here
    unsafe {
        *other = b'b'; // We can't even get a pointer to do this without invalidating `ptr`, and we
                       // can't do this write by casting `ptr` because it doesn't have write
                       // permission.
        dbg!(*ptr);
    }
}
```

**Comment by yaahc** [comment]
@Nilstrieb what happens when you apply an `#[unstable]` attribute to these methods? Is it just that it emits an unstable name collision warning lint?

**Comment by Nilstrieb** [comment]
```rust
error: an associated function with this name may be added to the standard library in the future
    --> compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:1499:25
     |
1499 |             vtable_name.as_ptr().cast(),
     |                         ^^^^^^
     |
     = warning: once this associated item is added to the standard library, the ambiguity may cause an error or change in behavior!
     = note: for more information, see issue #48919 <https://github.com/rust-lang/rust/issues/48919>
     = help: call with fully qualified syntax `bitflags::core::str::<impl str>::as_ptr(...)` to keep using the current method
     = help: add `#![feature(string_as_ptr)]` to the crate attributes to enable `std::string::String::as_ptr`
```

**Comment by RalfJung** [comment]
> (Elaborating on the "implied provenance handling": these and the underlying &str/&mut str methods would return the same values here, and questions about provenance handling differences between them have not been settled.)

(Nit: in terms of the Rust specification / abstract machine, they would *not* return the same values here, but values that differ in provenance. They just erase to the same bit pattern at runtime.)

**Comment by RalfJung** [comment]
As a possible argument not to have these methods, Tree Borrows makes this pretty much unnecessary -- the code in https://github.com/rust-lang/rust/issues/106593 is accepted with Tree Borrows.

**Comment by Nilstrieb** [comment]
It's probably also fair to say that this is something we _want_ to be allowed (at least I would want to :))

**Comment by RalfJung** [comment]
Now I worry about whether we can remove these inherent methods from `Vec` again...

**Comment by Nilstrieb** [comment]
That is probably lost hope. I guess we can say that they now exist because they are a tiny little less code than the deref path, making compilation faster :D

**Comment by RalfJung** [comment]
Hm, maybe we should open this again, or at least track this in an issue? There are still examples which are UB according to Tree Borrows that we might want to allow:
```rust
fn main() {
    let mut s = String::with_capacity(10);
    s.push('x');
    let ptr = s.as_mut_ptr();
    unsafe { ptr.write(0x20) };
    let _ptr2 = s.as_mut_ptr();
    unsafe { ptr.write(97) };
    println!("{s:?}");
}
```

**Pull request closed without merging**